compiler/hash-typecheck/src/traits.rs

```diff
@@ -185,10 +160,136 @@ pub struct CoreTraits {
+}
+
+impl<'c, 'w, 'm, 'ms, 'gs> TraitHelper<'c, 'w, 'm, 'ms, 'gs> {
+    pub fn new(
+        module_storage: &'ms mut ModuleStorage,
+        global_storage: &'gs mut GlobalStorage<'c, 'w, 'm>,
+    ) -> Self {
+        Self {
+            module_storage,
+            global_storage,
+        }
+    }
+
+    fn unifier(&mut self) -> Unifier<'c, 'w, 'm, '_, '_> {
+        Unifier::new(self.module_storage, self.global_storage)
+    }
+
+    pub fn find_trait_impl(
+        &mut self,
+        trt: &Trait,
+        trait_args: &[TypeId],
+        fn_type: Option<TypeId>,
+        trt_symbol: impl FnOnce() -> Symbol,
+        args_location: Option<SourceLocation>,
+    ) -> TypecheckResult<Substitution> {
+        let mut trait_args = trait_args.to_owned();
+        if trait_args.is_empty() {
+            trait_args.extend(
+                trt.args
+                    .iter()
+                    .map(|_| self.global_storage.types.create_unknown_type()),
+            )
+        }
+
+        if trait_args.len() != trt.args.len() {
+            return Err(TypecheckError::TypeArgumentLengthMismatch {
+                location: args_location.or_else(|| trt_symbol().location()),
+                expected: trt.args.len(),
+                got: trait_args.len(),
+            });
+        }
+
+        // Resolve any remaining fn args
+        let mut unifier = self.unifier();
+        if let Some(fn_type) = fn_type {
+            let trait_vars_sub = unifier.instantiate_vars_list(&trt.args)?;
+            let instantiated_fn = unifier.apply_sub(&trait_vars_sub, fn_type)?;
+            unifier.unify(fn_type, instantiated_fn, UnifyStrategy::ModifyBoth)?;
+        }
+
+        // let mut last_err = None;
+
+        // @@Performance: we have to collect due to lifetime issues, this is not ideal.
+        let impls: Vec<_> = self
+            .global_storage
+            .trait_impls
+            .get_impls(trt.id)
+            .iter()
+            .collect();
+        for (_, trait_impl) in impls.iter() {
+            match self.match_trait_impl(trait_impl, &trait_args) {
+                Ok(matched) => return Ok(matched),
+                Err(_e) => {
+                    continue;
+                    // last_err.replace(e);
+                }
+            }
+        }
+        // @@Todo: better errors
+        Err(TypecheckError::NoMatchingTraitImplementations(trt_symbol()))
+    }
+
+    pub fn print_types(&self, types: &[TypeId]) {
+        print!("[");
+        for &a in types {
+            print!("{}, ", TypeWithStorage::new(a, self.global_storage));
+        }
+        println!("]");
+    }
+
+    pub fn match_trait_impl(
+        &mut self,
+        trait_impl: &TraitImpl,
+        _trait_args: &[TypeId],
+    ) -> TypecheckResult<Substitution> {
+        let trt = self.global_storage.traits.get(trait_impl.trait_id);
+
+        // @@Ambiguity: for now let's assume all type variables in here are new
```

Comment:
Is this an issue with scoped type variables?

Reply:
No, this will actually change but for now I left it to instantiate all type variables. It is not really ambiguous, more like ambiguous to the programmer what it means. Like ```let foo<$T> = () => { let bar<$T> = () => { ..}; };` is the inner `$T` a new variable, or the same as the outer `$T`?

Reply:
Ah right ok